base/hashing2.jl

```diff
@@ -167,12 +167,35 @@ hash(x::Float16, h::UInt) = hash(float64(x), h)
 
 ## hashing collections ##
 const hashaa_seed = UInt === UInt64 ? 0x7f53e68ceb575e76 : 0xeb575e76
+const hashrle_seed = UInt == UInt64 ? 0x2aab8909bfea414c : 0xbfea414c
 function hash(a::AbstractArray, h::UInt)
     h += hashaa_seed
     h += hash(size(a))
-    for x in a
-        h = hash(x, h)
+
+    state = start(a)
+    done(a, state) && return h
+    x2, state = next(a, state)
+    done(a, state) && return hash(x2, h)
+
+    x1 = x2
+    while !done(a, state)
+        x1 = x2
+        x2, state = next(a, state)
+        if x2 == x1
+            # For repeated elements, use run length encoding
```

Comment:
No, `isequal` is not the natural way to distinguish different NaNs, that's what `===` is for (among other things). `isequal` is for hashing; we do not want all 2^53 different `NaN`s to hash differently. I guarantee you that having `d[NaN]` and `d[-NaN]` accessing different dict entries will cause no end of complaints.

This is very much outside of IEEE behavior, but then again, it's pretty clear that IEEE never thought for two seconds about hashing or other broader issues of equality in high-level programming languages. We have discussed making 0.0 and -0.0 hash the same, but I'm still not entirely sold on that.